### DISCUSSION TITLE
Feature Request: unwrapOrElse with default value

### DISCUSSION THREAD

**Discussion by DaviDevMod** [true-myth/true-myth#479] [Ideas]
What I am asking for is effectively to allow for `unwrapOrElse` to be an `unwrapOr` with side effects.

My use case is, having:

```ts
import Result, { ok, err, unwrapOrElse } from 'true-myth/result';

const fallibleNumber = (arg: boolean): Result<number, string> => (arg ? ok(5) : err('message'));

const fallibleString = (arg: boolean): Result<string, string> => (arg ? ok('yes') : err('message'));

const fallibleActiveElement = (): Result<Element, string> =>
  document.activeElement ? ok(document.activeElement) : err('message');
```

simplify this:
```ts
const n = fallibleNumber(false).unwrapOrElse((errValue: string) => {
  if (process.env.NODE_ENV !== 'production') throw new Error(errValue);

  return 0;
});

const s = fallibleString(false).unwrapOrElse((errValue: string) => {
  if (process.env.NODE_ENV !== 'production') throw new Error(errValue);

  return '';
});

const activeElement = fallibleActiveElement().unwrapOrElse((errValue: string) => {
  if (process.env.NODE_ENV !== 'production') throw new Error(errValue);

  return null;
});
```

to this:

```ts
const handleErr = <T>(errValue: string, defaultValue?: T) => {
  if (process.env.NODE_ENV !== 'production') throw new Error(errValue);

  return defaultValue;
};

const n = fallibleNumber(false).unwrapOrElse(handleErr, 0);

const s = fallibleString(false).unwrapOrElse(handleErr, '');

const activeElement = fallibleActiveElement().unwrapOrElse(handleErr, null);
```

What I did here was passing a default value as a second argument to `unwrapOrElse`, which is then somehow made available as an optional second argument for the `orElseFn` callback.

**Comment by chriskrycho** [comment]
Hmm. I see what you're going for here, but I don't think it actually makes sense to add this functionality to the library itself—you can imagine wanting to do the same with _every_ `orElse` function, for one thing; and for another it violates the standard contract for what an `orElse` function is (there actually is a contract for that in a type theory/category theory sense!).

That said, I think the thing you're trying to do can be solved pretty cleanly just using normal function composition. Instead of pushing it into True Myth, you should handle it with your own functions. That would look something like this:

```ts
function handleErr<T, E>(defaultValue: T): (err: E) => T {
  return (cause) => {
    if (process.env.NODE_ENV !== 'production') {
      throw new Error('Something went wrong', { cause });
    }

    return defaultValue;
  };
}
```

Then you just use that like this:

```ts
const n = fallibleNumber(false).unwrapOrElse(handleErr(0));
const s = fallibleString(false).unwrapOrElse(handleErr(''));
const activeElement = fallibleActiveElement()
    .unwrapOrElse(handleErr<Element | null, string>(null));
```

Here is a demo of that: [TS playground][p]. (Note the comment on the cast for the `activeElement` unwrapping: there's a reason for it!)

The nice thing here is that True Myth doesn't need to know anything about that pattern! *Any* function composition will work, whether that's doing side effects as here, or building it up via a pipeline of some *other* set of functions, or any other combination you can come up with. Instead of baking in some kind of specific handling for this one scenario, we have a fully general capability, just powered by function composition.

[p]: https://www.typescriptlang.org/play?target=99&moduleResolution=99&module=100#code/JYWwDg9gTgLgBAbzgUwB5mQYxgFQJ4YDyAZnAL5zFQQhwDkaG2AtDAcnQNwBQ3oksejCgBXZMxB4YACy59w0eACVkAZxEAbGABpEcCAGtdyKFF0iAdgHcoAQzCEoAUQ2rk5StVp1hYiVOkAeig1TRg5bkwIC1V4YlsNDWAAIw1kADkREGSTOABeOAAKWygAcwAuOGSICDTbCwBKSpV1LQAeCyycszhYqGALUoA+fKHuODgS0rgAfn0DQoBWBrhKkyhCuhA1VVtSjgaeSOjYygSk1OQAZWEB6YLissrq2uR6prgWsLa+u91fwYjPJjCZTWbzTZ4NR0FZrUybbaqXb7GFHKIxOLnFJpACC2GAADdkC5kNsLPAHh8vu0SWSdL1boDRuM4AATCCYLLIckAOls+KJtO58DmhkK7M5dL5AuJaTpsJQ8K2Oz2ByOxEs+OicGk9VZaScpjaOF0TiG4uQ8TCADUEmJKjgPoV1pUnCtgXAcIgWSEYCIoBYiphbCI3O6RggWRNgKRCmBqJgdjzuQSeelCAARJwAfSc6WtcAAhHkCnR4xBWSItRYYd6JvW4DJqFY4BZkC3DdQNnQrjRkDI7nArMKh9RBnRdEhg6H3GRDlHyLwG77-YHWZaQ1pbRoxDwJmQeGReOjToGCvFEtiMl0TIUL2GeZYbPZHC43IVdRZ9cT4QAGBrzie8CqPkZyXpcNz9IMd4JA+T52A4ziuMgH56gaSowoBJzwPyMCErKpIjueWKXHieGCnKwqFA0j7WAhr7IYULKBIEnrSO4VjAOuUCNuwcDACBIQAI4iMAISsjqJjuMkgSYJ6VxwMGNbwJYMZ4HAAAGQrkhpkxfppnSJBpzGsZxMgQCI8DADAPJse4CQwCYFi2ORyC6HqcAgFW0hwPGlrrMgEn2OW-LSLoVgWRoEk5I2EAmXAABUFgQDACVwCuAYGZoGi6exIQ8iyn7fp2bTafAAA+rbZf8jLDIUhkaA03DztwjBYLg7AkPVNEwBAABC17ZLe85tdg+BEMQhSqD1-XXLV1E8KNHUTcUMplfOQA

**Comment by DaviDevMod** [comment]
@chriskrycho you are right on everything.
I should have thought about just returning a function.
Thank you so much for the thorough reply.

**Reply by chriskrycho** [reply]
No problem; happy to help!
